tiledb/sm/query/ast/query_ast.cc

```diff
@@ -0,0 +1,306 @@
+}
+
+/** Value node getter methods */
+const std::string& ASTNodeVal::get_node_field_name() const {
+  return field_name_;
+}
+const ByteVecValue& ASTNodeVal::get_node_condition_value_data() const {
+  return condition_value_data_;
+}
+const UntypedDatumView& ASTNodeVal::get_node_condition_value_view() const {
+  return condition_value_view_;
+}
+const QueryConditionOp& ASTNodeVal::get_node_op() const {
+  return op_;
+}
+
+/** Expression node getter methods */
+const std::vector<tdb_unique_ptr<ASTNode>>& ASTNodeVal::get_node_children()
+    const {
+  throw std::runtime_error(
+      "ASTNodeVal::get_node_children: called get_node_children() on a value "
+      "node.");
+}
+const QueryConditionCombinationOp& ASTNodeVal::get_node_combination_op() const {
+  throw std::runtime_error(
+      "ASTNodeVal::get_node_combination_op: called get_node_combination_op on "
+      "a value node.");
+}
+
+/** Returns true is the node is an expression node. */
+bool ASTNodeExpr::is_expr() const {
+  return true;
+}
+
+/** Returns a deep copy of the node. */
+tdb_unique_ptr<ASTNode> ASTNodeExpr::clone() const {
+  std::vector<tdb_unique_ptr<ASTNode>> nodes_copy;
+  for (const auto& node : nodes_) {
+    nodes_copy.push_back(node->clone());
+  }
+  return tdb_unique_ptr<ASTNode>(
+      tdb_new(ASTNodeExpr, std::move(nodes_copy), combination_op_));
+}
+
+void ASTNodeExpr::get_field_names(
+    std::unordered_set<std::string>& field_name_set) const {
+  for (const auto& child : nodes_) {
+    child->get_field_names(field_name_set);
+  }
+}
+
+bool ASTNodeExpr::is_previously_supported() const {
+  if (combination_op_ != QueryConditionCombinationOp::AND) {
+    return false;
+  }
+  for (const auto& child : nodes_) {
+    if (child->is_expr()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+Status ASTNodeExpr::check(const ArraySchema& array_schema) const {
+  // If the node is a compound expression node, ensure there are at least
+  // two children in the node and then run a check on each child node.
+  if (nodes_.size() < 2) {
+    return Status_QueryConditionError(
+        "Non value AST node does not have at least 2 children.");
+  }
+  for (const auto& child : nodes_) {
+    RETURN_NOT_OK(child->check(array_schema));
+  }
+
+  return Status::Ok();
+}
+
+tdb_unique_ptr<ASTNode> ASTNodeExpr::combine(
+    const tdb_unique_ptr<ASTNode>& rhs,
+    const QueryConditionCombinationOp& combination_op) {
+  std::vector<tdb_unique_ptr<ASTNode>> ast_nodes;
+  if (!rhs->is_expr()) {
+    if (combination_op == combination_op_) {
+      for (const auto& child : nodes_) {
+        ast_nodes.push_back(child->clone());
+      }
+    } else {
+      ast_nodes.push_back(this->clone());
+    }
+    ast_nodes.push_back(rhs->clone());
+  } else {
+    if (combination_op == combination_op_) {
+      for (const auto& child : nodes_) {
+        ast_nodes.push_back(child->clone());
+      }
+    } else {
+      ast_nodes.push_back(this->clone());
+    }
+
```

Comment:
This much of the `if (!rhs->is_expr()` and the `else` seem almost identical.  Maybe replace with just 
```c++
  if (combination_op == combination_op_) {
     for (const auto& child : nodes_) {
       ast_nodes.push_back(child->clone());
     }
   } else {
     ast_nodes.push_back(this->clone());
   }
   if (!rhs->is_expr) {
     ast_nodes.push_back(rhs->clone());
   } else {
     ...
   }
```
(You might be able to simplify further -- there is a duplicated 
```c++
      ast_nodes.push_back(rhs->clone());
``` 
as well.

Reply:
Simplified this!